stratum/hal/lib/tdi/tdi_id_mapper.cc

```diff
@@ -0,0 +1,285 @@
+  // called "[pipeline name].[P4 info table name]"
+  std::vector<const tdi::Table*> tdi_tables;
+  RETURN_IF_TDI_ERROR(tdi_info->tablesGet(&tdi_tables));
+  for (const auto* table : tdi_tables) {
+    tdi_id_t table_id;
+    std::string table_name;
+    table_id = table->tableInfoGet()->idGet();
+    table_name = table->tableInfoGet()->nameGet();
+    if (absl::StrContains(table_name, p4info_name)) {
+      p4info_to_tdi_id_[p4info_id] = table_id;
+      tdi_to_p4info_id_[table_id] = p4info_id;
+      return ::util::OkStatus();
+    }
+  }
+  return MAKE_ERROR(ERR_INTERNAL)
+         << "Unable to find tdirt ID for P4Info entity " << p4info_name
+         << " with ID " << p4info_id << ".";
+}
+
+::util::Status TdiIdMapper::BuildActionProfileMapping(
+    const p4::config::v1::P4Info& p4info, const tdi::TdiInfo* tdi_info,
+    const std::string& context_json_content) {
+  absl::flat_hash_map<std::string, std::string> prof_to_sel;
+  try {
+    nlohmann::json context_json =
+        nlohmann::json::parse(context_json_content, nullptr, false);
+    CHECK_RETURN_IF_FALSE(!context_json.is_discarded())
+        << "Failed to parse context.json";
+
+    // Builds mappings for ActionProfile and ActionSelector.
+    for (const auto& table : context_json["tables"]) {
+      if (!table.contains("action_profile")) {
+        continue;
+      }
+      const auto& action_profile_name = table["action_profile"];
+      if (action_profile_name.empty()) {
+        // Skip the table if there is no ActionProfile supported.
+        continue;
+      }
+
+      if (!table.contains("selection_table_refs")) {
+        continue;
+      }
+      const auto& selection_table_refs = table["selection_table_refs"];
+      if (selection_table_refs.empty()) {
+        // Skip the table if it supports ActionProfile only, since we don't need
+        // to create the mapping for this table.
+        continue;
+      }
+
+      const auto& action_selector_name = selection_table_refs[0]["name"];
+      CHECK_RETURN_IF_FALSE(!action_selector_name.empty())
+          << "ActionSelector for ActionProfile " << action_profile_name
+          << " name is empty, this should not happened";
+      CHECK_RETURN_IF_FALSE(gtl::InsertIfNotPresent(
+          &prof_to_sel, action_profile_name, action_selector_name))
+          << "Action profile with name " << action_profile_name
+          << " already exists.";
+    }
+  } catch (nlohmann::json::exception& e) {
+    return MAKE_ERROR(ERR_INTERNAL) << e.what();
+  }
+
+  // Searching all action profile and selector tables from tdirt.json
+  absl::flat_hash_map<std::string, tdi_id_t> act_prof_tdi_ids;
+  absl::flat_hash_map<std::string, tdi_id_t> selector_tdi_ids;
+  std::vector<const tdi::Table*> tdi_tables;
+  RETURN_IF_TDI_ERROR(tdi_info->tablesGet(&tdi_tables));
+  for (const auto* table : tdi_tables) {
+    std::string table_name;
+    tdi_id_t table_id;
+    auto table_type = static_cast<tdi_rt_table_type_e>(
```

Comment:
tdi_rt_table_type_e and TDI_RT_TABLE_TYPE_ACTION_PROFILE are DPDK specific.
We need to get the relevant folks to figure this out.
This also requires a P4OVS_CHANGES.

Reply:
Same comment for lines 203 and 207.

Reply:
This is wrong. The set of possible table types should be common; otherwise, TDI is not abstracting the underlying driver.